[Inductor] make decompose_mm_pass support cpu case

hl475 opened the pull request:
Summary: Previously, decompose_mm_pass only works for gpu case. This diff make it support some cpu case as well for the performance optimization

Differential Revision: D65226131




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

hl475 commented:
@pytorchbot label "topic: not user facing"

eellison commented:
I believe @Chillee has done benchmarking and this is not consistently profitable unless coordinate descent tuning. Also this is very similar to `decompose_mem_bound_mm` which isn't on by default. cc @mengluy0125

eellison commented:
Why is decompose_mem_bound_mm not usable here ? what's the difference

hl475 commented:
I can reuse decompose_mem_bound_mm for this case as well!

hl475 commented:
I thought you wanted me to follow the same approach to create another config, but if you think I can just use decompose_mem_bound_mm, then yes, I can use that!

eellison commented:
@hl475 have you tried enablign the config on your model without makign changes? how does that go ?

hl475 commented:
@eellison will try it out and let you know! 

To confirm - you want to see what would be the performance if I just change config.decompose_mem_bound_mm=True, but no change on decompoistion at all, right?

eellison commented:
That's correct.

eellison commented on torch/_inductor/fx_passes/decompose_mem_bound_mm.py:
```diff
@@ -29,8 +29,12 @@
 
-def check_device(a: Tensor, b: Tensor) -> bool:
-    return a.is_cuda and b.is_cuda
+def check_device(a: Tensor, b: Tensor, device="cuda") -> bool:
+    if device == "cuda":
+        return a.is_cuda and b.is_cuda
+    elif device == "cpu":
+        return a.is_cpu and b.is_cpu
+    return False
```
this is just (a.device.type == b.device.type) and (b.device.type == device)

eellison commented on torch/_inductor/fx_passes/decompose_mem_bound_mm.py:
```diff
@@ -57,7 +59,13 @@ def should_decompose_bmm(mat1, mat2) -> bool:
-    return True
+        return True
+    if check_device(mat1, mat2, device="cpu"):
+        if len(mat1.shape) != 3 or len(mat2.shape) != 3:
```
We can factor this above tboth `check_device` if branches right ?

eellison commented on torch/_inductor/fx_passes/decompose_mem_bound_mm.py:
```diff
@@ -57,7 +59,13 @@ def should_decompose_bmm(mat1, mat2) -> bool:
+    if check_device(mat1, mat2, device="cpu"):
+        if len(mat1.shape) != 3 or len(mat2.shape) != 3:
+            return False
+        if mat1.shape[0] == 1 and mat2.shape[0] == 1:
+            return True
```
I know it's already being done this way in the pass but nonetheless could we use guard_shape_oblivious for the new checks youve added ?

eellison commented on torch/_inductor/fx_passes/decompose_mem_bound_mm.py:
```diff
@@ -73,6 +81,13 @@ def should_decompose_mm(mat1, mat2) -> bool:
         and mat2.shape[1] < max_other_dimention_decomposition
+    ) or (
+        check_device(mat1, mat2, device="cpu")
+        and len(mat1.shape) == 2
```
Similarly, these checks can be factored above for both devices

eellison reviewed (approved):
looks good a few nits

hl475 commented:
address comments

hl475 committed:
[Inductor] make decompose_mm_pass support cpu case (#139696)

Summary:

Previously, decompose_mm_pass only works for gpu case. This diff make it support some cpu case as well for the performance optimization

Differential Revision: D65226131

hl475 commented:
lint failure on test/test_nestedtensor.py seems pre-existing

hl475 commented:
@pytorchbot merge

hl475 commented:
@pytorchbot merge -f "lint failure on test/test_nestedtensor.py seems pre-existing"

etaf commented on test/inductor/test_decompose_mem_bound_mm.py:
```diff
@@ -117,6 +118,29 @@ def test_decompose_bmm(self, b, m, n, k, should_decompose):
         )
         counters.clear()
 
+    @parametrize(
```
Hi, @hl475  @eellison  actually this case is not tested because the the test class is decorated as @requires_gpu, and if I run this case in machine that has gpu, it fails. So I suggest refine the case.
